### PR TITLE
[WIP] Add Storage Transformers to Core Spec & Sharding v1.0

### PR DESCRIPTION
**Pull request by jstriebel**
To get an auto-reloading server, run (using venv)
```
python3 -m venv .venv
. .venv/bin/activate
pip install -r docs/requirements
pip install sphinx-autobuild
sphinx-autobuild docs docs/_build/html
```

Adapting the abstract store interface is still TODO, it needs support for partial and bundled access for multiple chunks.

Some open questions remain, e.g.
* Should we introduce the abstract concept of storage transformers? I think this would be helpful for many other use cases.
* Storage transformers vs extensions:
  * Should storage transformers have a `must_understand` flag?
  * Can storage transformers be extensions?
  * Are all of them extensions?
  * Should sharding be defined as an extension?
  
Those could be up for discussion in the final PR.

### CODE REVIEW

File: docs/storage_transformers/sharding/v1.0.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+with `data/root`, as they indicate data keys for array chunks. All other
+entries are simply passed on.
+
+Entries starting with ``data/root`` are grouped by their common shard, assuming
+storage keys from a regular chunk grid which may use a customly configured
+``chunk separator``:
+For all entries that are part of the same shard the key is changed to the
+shard-key and the values are combined in the `Binary shard format`_ described
+below. The new shard-key is the chunk key divided by ``chunks_per_shard`` and
+floored per dimension. E.g. for ``chunks_per_shard=[32, 2]``, the chunk grid
```
Comment (valentin-pinkau) [inline docs/storage_transformers/sharding/v1.0.rst]: Maybe start the example similar as above with: For example...

File: docs/storage_transformers/sharding/v1.0.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    for each dimension of the Zarr array, where each chunk may only start at a position
+    that is divisble by ``chunks_per_shard`` per dimension, e.g. starting at the zero-origin.
+    The length of the array must match the length of the array metadata ``shape`` entry.
+    For example, a value ``[32, 2]`` indicates that 64 chunks are combined in one shard,
+    32 along the first dimension, and for each of those 2 along the second dimension.
+    Valid starting positions for a shard in the chunk-grid are therefore ``[0, 0]``,
+    ``[32, 2]``, ``[32, 4]``, ``[64, 2]`` or ``[96, 18]``.
+
+
+Key & value transformation
```
Comment (valentin-pinkau) [inline docs/storage_transformers/sharding/v1.0.rst:110]: This is basically the storage transformer implementation for sharding, right?
Maybe reflect that in this heading.

Comment (valentin-pinkau) [reply docs/storage_transformers/sharding/v1.0.rst:110]: Key & value transformation with sharding storage transformer?

File: docs/storage_transformers/sharding/v1.0.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+* **Fixed order**: Specify a fixed order (e.g. row-, column-major or Morton order).
+  When replacing existing chunks larger or equal sized chunks may be replaced in-place,
+  leaving unused space up to an upper limit  which might possibly be specified.
+  Please note that for regular-sized uncompressed data all chunks have the same size and
+  can therefore be replaced in-place.
+* **Append-only**: Any chunk to write is appended to the existing shard, followed by an updated index.
+
+Any configuration parameters for the write strategy must not be part of the metadata document,
+in a shard I'd propose to use Morton order, but this can easily be changed and customized, since any order can be read.
+
```
Comment (valentin-pinkau) [inline docs/storage_transformers/sharding/v1.0.rst]: This sounds like an opinion, which should not be part of the spec, what do you think?

File: docs/storage_transformers/sharding/v1.0.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+Binary shard format
+===================
+
+The only binary format is the ``indexed`` format, as specified by the ``format``
+configuration key. Other binary formats might be added in future versions.
+
+In the indexed binary format chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them.
+The index holds an `offset, length` pair of little-endian uint64 per chunk,
```
Comment (valentin-pinkau) [inline docs/storage_transformers/sharding/v1.0.rst:148]: At which offset do I find the index?

**Review by valentin-pinkau — COMMENTED** [review]
Very nice! I left some comments :)

File: docs/storage_transformers/sharding/v1.0.rst (hunk 3 above)
Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst]: Good point, that was just a left-over from a copy and paste, removed it and adapted the previous sentence

File: docs/storage_transformers/sharding/v1.0.rst (hunk 4 above)
Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst:148]: At the end of the file, I added an explanation.

File: docs/storage_transformers/sharding/v1.0.rst (hunk 2 above)
Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst:110]: Instead I added `Storage transformer implementation` as another heading and put the other two as subsections:

```
Storage transformer implementation
==================================

Key & value transformation
--------------------------

…


Binary shard format
-------------------

…
```

**Comment by jstriebel** [comment]
@valentin-pinkau Danke dir für's Feedback! Kannst du vielleicht auch nochmal über den letzten Commit 28a9023 schauen, ob das für dich so Sinn ergibt?

File: docs/storage_transformers/sharding/v1.0.rst [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+========
+
+This specification defines an implementation of the Zarr
+storage transformer protocol for sharding.
+
+
+Motivation
+==========
+
+Sharding decouples the concept of chunks from storage keys, which become shards.
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]: That is not entirely true. Shards hold 0 or more chunks, but not partial chunks.

A definition of sharding could be "co-locating multiple chunks within a file/storage object"

File: docs/storage_transformers/sharding/v1.0.rst [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+This is helpful when the requirements for those don't align:
+
+- Compressible units of chunks often need to be read and written in smaller
+  chunks, whereas
+- storage often is optimized for larger data per entry and fewer entries, e.g.
+  as restricted by the file block size and maximum inode number for typical
+  file systems.
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
- it becomes inefficient or impractical to store a large number of chunks in single files or objects due to the design constraints of the underlying storage, e.g. as restricted by the file block size and maximum inode number for typical file systems.
```

File: docs/storage_transformers/sharding/v1.0.rst [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+
+Motivation
+==========
+
+Sharding decouples the concept of chunks from storage keys, which become shards.
+This is helpful when the requirements for those don't align:
+
+- Compressible units of chunks often need to be read and written in smaller
+  chunks, whereas
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
- Chunk sizes need to be small for read efficiency requirements, e.g. for data streaming in browser-based visualization software, whereas
```

writing is not really facilitated through sharding.

File: docs/storage_transformers/sharding/v1.0.rst [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+- storage often is optimized for larger data per entry and fewer entries, e.g.
+  as restricted by the file block size and maximum inode number for typical
+  file systems.
+
+This does not necessarily fit the access patterns of the data, so chunks might
+need to be smaller than one storage key. In those cases sharding decouples those
+entities. One shard corresponds to one storage key, but can contain multiple chunks:
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
This does not necessarily fit the access patterns of the data, so chunks might
need to be smaller than the minimum size of one storage key. In those cases sharding decouples those
entities. One shard corresponds to one storage key, but can contain multiple chunks:
```

File: docs/storage_transformers/sharding/v1.0.rst [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+Storage transformer implementation
+==================================
+
+Key & value transformation
+--------------------------
+
+The storage transformer protocol defines the abstract interface to be the same
+as the :ref:`abstract-store-interface`.
+
+The Zarr store interface is defined in terms of `keys` and `values`,
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
The Zarr store interface is defined as a mapping of `keys` and `values`,
```

File: docs/storage_transformers/sharding/v1.0.rst [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+where a `key` is a sequence of characters and a `value` is a sequence
+of bytes. A key-value pair is called `entry` in the following part.
+
+This sharding transformer only adapts entries where the key starts
+with `data/root`, as they indicate data keys for array chunks. All other
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]: where does that come from? is that a constant or an example? can this be referenced to another part of the spec?

File: docs/storage_transformers/sharding/v1.0.rst [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+Binary shard format
+-------------------
+
+The only binary format is the ``indexed`` format, as specified by the ``format``
+configuration key. Other binary formats might be added in future versions.
+
+In the indexed binary format chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them.
+The index is placed at the end of the file and has a length of 16 bytes per chunk
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
The index is placed at the end of the file and has a length of 16 bytes multiplied by the number of chunks
```

File: docs/storage_transformers/sharding/v1.0.rst [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+  When replacing existing chunks larger or equal sized chunks may be replaced in-place,
+  leaving unused space up to an upper limit  which might possibly be specified.
+  Please note that for regular-sized uncompressed data all chunks have the same size and
+  can therefore be replaced in-place.
+* **Append-only**: Any chunk to write is appended to the existing shard,
+  followed by an updated index.
+
+Any configuration parameters for the write strategy must not be part of the metadata document,
+they need to be configured at runtime, as this is implementation specific.
+
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst:178]: I think what is missing is a read and write protocol, i.e. how does reading/writing work in an abstract algorithm.

**Review by normanrz — COMMENTED** [review]
Most of it reads great!
- I think a read/write protocol is missing from the sharing spec.
- The "Motivation" doesn't sound convincing/concise enough to me. I added some suggestions, not sure if they help 😅

File: docs/storage_transformers/sharding/v1.0.rst (hunk 10 above)
Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst]: That's part of the storage-key spec, a constant. I added a reference

**Comment by jstriebel** [comment]
Thanks for the great feedback!
> I think a read/write protocol is missing from the sharing spec.

I added an `API implementation` chapter now covering this.

> The "Motivation" doesn't sound convincing/concise enough to me. I added some suggestions, not sure if they help sweat_smile

Yep, I rewrote it quite a bit again, I personally think it's better now. I'm not even sure if it should be part of the spec at all, but I think it's still helpful. What do you think?

Could you please give those two parts "Motivation" & "API implementation" in `docs/storage_transformers/sharding/v1.0.rst` another look @normanrz & @valentin-pinkau? Thanks a ton :pray:

File: docs/storage_transformers/sharding/v1.0.rst [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+In many cases it becomes inefficient or impractical to store a large number of chunks in
+single files or objects due to the design constraints of the underlying storage,
+for example as restricted by the file block size and maximum inode number for typical file systems.
+
+Increasing the chunk size works only up to a certain point, as chunk sizes need to be small for
+read efficiency requirements, for example to stream data in browser-based visualization software.
+
+Therefore, chunks may need to be smaller than the minimum size of one storage key.
+In those cases it is required to store objects at a more coarse granularity than reading chunks.
```
Comment (normanrz) [inline docs/storage_transformers/sharding/v1.0.rst]:
```suggestion
In those cases it is efficient to store objects at a more coarse granularity than reading chunks.
```

**Review by normanrz — APPROVED** [review]
I think this is good for sending out for public feedback!

> I'm not even sure if it should be part of the spec at all, but I think it's still helpful. What do you think?

I think this should definitely be in the spec in order to convince people that sharding is important to implement.

The `API implementation` could even be a bit more high-level for my taste, but let's see what the feedback is.

File: docs/storage_transformers/sharding/v1.0.rst [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+  This operation should be implemented using two ``get_partial_values``
+  operations on the underlying store, one for retrieving the indices and
+  one for retrieving existing chunks.
+
+* ``set_partial_values(key_start_values)`` :
+  For each referenced key, check if all available chunks in a shard are
+  referenced. In this case a shard can be constructed according to the
+  `Binary shard format`_ directly.
+  For all other keys, request the indices from the underlying store using
+  ``get_partial_values``. All chunks that are not updated completely and
```
Comment (valentin-pinkau) [inline docs/storage_transformers/sharding/v1.0.rst:206]: how does the `get_partial_values` function return the indices?

**Review by valentin-pinkau — COMMENTED** [review]
Very nice, I have one question:

File: docs/storage_transformers/sharding/v1.0.rst (hunk 14 above)
Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst:206]: The index format is described above.

Comment (valentin-pinkau) [reply docs/storage_transformers/sharding/v1.0.rst:206]: So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?
To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Comment (valentin-pinkau) [reply docs/storage_transformers/sharding/v1.0.rst:206]: But I guess it makes sense :)

Comment (jstriebel) [reply docs/storage_transformers/sharding/v1.0.rst:206]: > So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?

Exactly :+1: 

> To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Yes, a bit, but that's also part of the beauty, as nobody cares if the "next" store is actually the store or just the next transformer. I tried to always use the term `underlying store` refering to the next transformer or store, as described in the paragraph above in the spec:
> The term `underlying store` references either the next storage transformer
in the stack or the actual store if this transformer is the last one in the

However, it's still confusing, but I have no idea how to improve it. Do you have a suggestion?

**Pull request closed without merging**